File: src/contests/cg/puzzle/vbnet/max-area.vb
' 🎮 CodinGame Puzzle - max-area
' https://www.codingame.com/training/easy/max-area

Imports System

Module Solution
    Sub Main()
        Console.ReadLine()
        Dim tokens() As String = Console.ReadLine().Trim().Split(" "c)
        Dim a(tokens.Length - 1) As Long
        For i As Integer = 0 To tokens.Length - 1
            a(i) = Long.Parse(tokens(i))
        Next
        Dim left As Integer = 0, right As Integer = a.Length - 1
        Dim best As Long = 0
        While left < right
            Dim h As Long = Math.Min(a(left), a(right))
            Dim area As Long = h * (right - left)
            If area > best Then best = area
            If a(left) < a(right) Then
                left += 1
            Else
                right -= 1
            End If
        End While
        Console.WriteLine(best)
    End Sub
End Module

File: src/contests/cg/puzzle/haskell/max-area.hs
-- 🎮 CodinGame Puzzle - max-area
-- https://www.codingame.com/training/easy/max-area

import Data.Array

main :: IO ()
main = do
  _ <- getLine
  line <- getLine
  let xs = map read (words line) :: [Integer]
      n = length xs
      arr = listArray (0, n - 1) xs
      go l r best
        | l >= r = best
        | otherwise =
            let h = min (arr ! l) (arr ! r)
                area = h * fromIntegral (r - l)
                nb = max best area
             in if arr ! l < arr ! r then go (l + 1) r nb else go l (r - 1) nb
  print (go 0 (n - 1) 0)
